### PR TITLE
Build coverage information from abitest run

### PR DESCRIPTION
**Pull request by daviddrysdale**
# Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by
        [Cloudbuild](/cloudbuild.yaml)
  - [ ] I have updated [documentation](/docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

**Comment by daviddrysdale** [comment]
Initial coverage info at https://codecov.io/gh/project-oak/oak/branch/coverage

**Comment by daviddrysdale** [comment]
The `LCOV_EXCL_{START,STOP}` markers don't seem to work with codecov.io, but they do work with local tools (e.g. lcov's `genhtml`) so I think the second commit is probably worth keeping (in the hope that codecov.io acquires support for range exclusions in the future).

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs [hunk 1]
```diff
@@ -536,8 +536,8 @@ impl wasmi::Externals for WasmInterface {
             WAIT_ON_CHANNELS => {
                 map_host_errors(self.wait_on_channels(args.nth_checked(0)?, args.nth_checked(1)?))
             }
-            WASI_STUB => panic!("Attempt to invoke unimplemented WASI function"),
-            _ => panic!("Unimplemented function at {}", index),
+            WASI_STUB => panic!("Attempt to invoke unimplemented WASI function"), // LCOV_EXCL_LINE
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/wasm/mod.rs:539]: Are we expected to annotate code in this way going forward? If so, all panics, or just some?

File: .github/workflows/upload_coverage.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        run: |
+          docker pull gcr.io/oak-ci/oak:latest
+          docker build --pull --cache-from=gcr.io/oak-ci/oak:latest --tag=gcr.io/oak-ci/oak:latest .
+
+      - name: Generate coverage data
+        run: ./scripts/docker_run ./scripts/run_tests_coverage
+
+      - name: Upload to codecov.io
+        run: |
+          bash <(curl -s https://codecov.io/bash) -X gcov -f ./oak/server/target/debug/lcov.info
```
Comment (tiziano88) [inline .github/workflows/upload_coverage.yaml:38]: How does this authenticate (if at all) to codecov.io?

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs (hunk 1 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/wasm/mod.rs]: Yeah, that would be a pain &ndash; I've dropped the manual exclusion, and set `panic!` and `unreachable!` to count as unhittable.

File: .github/workflows/upload_coverage.yaml (hunk 2 above)
Comment (daviddrysdale) [reply .github/workflows/upload_coverage.yaml]: There seems to be some sort of check from codecov.io back to the GitHub actions server, to allow [tokenless uploads](https://blog.codecov.io/2020/03/09/tokenless-uploads-for-github-actions/), but I'm not sure of the details.

Comment (tiziano88) [reply .github/workflows/upload_coverage.yaml]: I could imagine that working if using [their own action](https://github.com/codecov/codecov-action) , but I have no idea how this bash one liner actually pulls it off. Not sure if I should be happy or scared about that.

**Comment by tiziano88** [comment]
Where do we see the output? Perhaps add a link / badge to the readme?

File: scripts/run_tests_coverage [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    "--private_key=${SCRIPTS_DIR}/../examples/certs/local/local.key")
+readonly CLIENT_ARGS=("${@-}")  # Choose client args provided after '--'.
+
+# Run the client program.
+"${SCRIPTS_DIR}/../bazel-client-bin/examples/abitest/client/client" "${TLS_ARGS[@]}" "${ADDITIONAL_ARGS[@]-}" "${CLIENT_ARGS[@]-}"
+
+# Shutdown cleanly.
+kill_bg_pids
+
+# Generate and upload coverage summary
```
Comment (tiziano88) [inline scripts/run_tests_coverage:44]: Is this actually uploading it too?

File: .github/workflows/upload_coverage.yaml (hunk 2 above)
Comment (daviddrysdale) [reply .github/workflows/upload_coverage.yaml]: The bash one liner pulls in a 1750-line script, so probably scared&hellip;

**Pull request merged**